File: src/core/objects/tree/tree-entry.ts
export enum EntryType {
  DIRECTORY = '040000',
  REGULAR_FILE = '100644',
  EXECUTABLE_FILE = '100755',
  SYMBOLIC_LINK = '120000',
  SUBMODULE = '160000',
}

/**
 * Represents a single entry in a Git tree object.
 *
 * Each entry contains:
 * - mode: File permissions and type (6 bytes, octal)
 * - name: Filename or directory name (variable length string)
 * - sha: SHA-1 hash of the referenced object (40 character hex string)
 *
 * Entry types by mode:
 * - 040000: Directory (tree object)
 * - 100644: Regular file (blob object)
 * - 100755: Executable file (blob object)
 * - 120000: Symbolic link (blob object)
 * - 160000: Git submodule (commit object)
 *
 * Serialized format in tree object:
 * [mode] [space] [filename] [null byte] [20-byte SHA-1 binary]
 *
 * Example serialized entry for "hello.txt" file:
 * "100644 hello.txt\0[20 bytes of SHA-1]"
 */

export class TreeEntry {
  private _mode: string;
  private _name: string;
  private _sha: string;

  private static readonly NULL_BYTE = 0x00;
  private static readonly SPACE_BYTE = 0x20;
  private static readonly SHA_LENGTH_BYTES = 20;

  constructor(mode: string, name: string, sha: string) {
    this._mode = mode;
    this._name = this.validateName(name);
    this._sha = this.validateSha(sha);
  }

  mode(): string {
    return this._mode;
  }

  name(): string {
    return this._name;
  }

  sha(): string {
    return this._sha;
  }

  entryType(): EntryType {
    return TreeEntry.fromMode(this._mode);
  }

  static fromMode(mode: string): EntryType {
    const entryType = Object.values(EntryType).find((type) => type === mode);
    if (!entryType) {
      throw new Error(`Unknown mode: ${mode}`);
    }
    return entryType;
  }

  isDirectory(): boolean {
    return this.entryType() === EntryType.DIRECTORY;
  }

  isFile(): boolean {
    const type = this.entryType();
    return type === EntryType.REGULAR_FILE || type === EntryType.EXECUTABLE_FILE;
  }

  isExecutable(): boolean {
    return this.entryType() === EntryType.EXECUTABLE_FILE;
  }

  isSymbolicLink(): boolean {
    return this.entryType() === EntryType.SYMBOLIC_LINK;
  }

  isSubmodule(): boolean {
    return this.entryType() === EntryType.SUBMODULE;
  }

  /**
   * Serialize this entry for inclusion in a tree object
   * Format: [mode] [space] [filename] [null byte] [20-byte SHA-1 binary]
   */
  serialize(): Uint8Array {
    const modeAndName = `${this.mode} ${this.name}\0`;
    const modeNameBytes = new TextEncoder().encode(modeAndName);

    // Convert SHA-1 hex string to binary (20 bytes)
    const shaBytes = new Uint8Array(20);
    for (let i = 0; i < 20; i++) {
      const index = i * 2;
      shaBytes[i] = parseInt(this._sha.substring(index, index + 2), 16);
    }

    const result = new Uint8Array(modeNameBytes.length + shaBytes.length);
    result.set(modeNameBytes, 0);
    result.set(shaBytes, modeNameBytes.length);

    return result;
  }

  /**
   * Compare this entry with another entry.
   * The comparison is based on the entry's name, with directories sorted before files.
   * @param other - The other entry to compare with.
   * @returns A negative value if this entry is less than the other, zero if they are equal, or a positive value if this entry is greater than the other.
   */
  compareTo(other: TreeEntry): number {
    const thisKey = this.isDirectory() ? this._name + '/' : this._name;
    const otherKey = other.isDirectory() ? other._name + '/' : other._name;

    if (thisKey < otherKey) {
      return -1;
    }
    if (thisKey > otherKey) {
      return 1;
    }
    return 0;
  }

  /**
   * Create a TreeEntry from serialized data
   */
  static deserialize(data: Uint8Array, offset: number): { entry: TreeEntry; nextOffset: number } {
    let spaceIndex = offset;
    while (spaceIndex < data.length && data[spaceIndex] !== TreeEntry.SPACE_BYTE) {
      spaceIndex++;
    }

    const mode = new TextDecoder().decode(data.slice(offset, spaceIndex));

    let nullIndex = spaceIndex + 1;
    while (nullIndex < data.length && data[nullIndex] !== TreeEntry.NULL_BYTE) {
      nullIndex++;
    }

    const name = new TextDecoder().decode(data.slice(spaceIndex + 1, nullIndex));

    const shaStartIndex = nullIndex + 1;
    const shaEndIndex = shaStartIndex + TreeEntry.SHA_LENGTH_BYTES;
    const shaBytes = data.slice(shaStartIndex, shaEndIndex);
    const sha = Array.from(shaBytes)
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');

    return {
      entry: new TreeEntry(mode, name, sha),
      nextOffset: nullIndex + TreeEntry.SHA_LENGTH_BYTES + 1,
    };
  }

  /**
   * Validate the name of the entry.
   * Git doesn't allow certain characters in filenames
   */
  private validateName(name: string): string {
    if (name == null || name.length === 0) {
      throw new Error('Name cannot be null or empty');
    }

    // Git doesn't allow certain characters in filenames
    if (name.includes('/') || name.includes('\0')) {
      throw new Error(`Invalid characters in name: ${name}`);
    }

    return name;
  }

  /**
   * Validate the SHA of the entry.
   * The SHA should only be hex characters and should not exceed 40 chars
   */
  private validateSha(sha: string): string {
    const expectedLength = TreeEntry.SHA_LENGTH_BYTES * 2;
    if (sha == null || sha.length !== expectedLength) {
      throw new Error(`SHA must be ${expectedLength} characters long`);
    }

    // Validate hex characters
    if (!sha.match(/[0-9a-fA-F]+/)) {
      throw new Error('SHA must contain only hex characters');
    }

    return sha.toLowerCase();
  }
}

File: sourcecontrol/pkg/objects/tree/tree_entry.go
package tree

import (
	"bytes"
	"encoding/hex"
	"fmt"
	"io"

	"github.com/utkarsh5026/SourceControl/pkg/objects"
	"github.com/utkarsh5026/SourceControl/pkg/repository/scpath"
)

const (
	SHALengthBytes = 20
)

// TreeEntry represents a single entry in a Git tree object.
//
// Each entry contains:
// - mode: File permissions and type (FileMode)
// - name: Filename or directory name (RelativePath)
// - sha: SHA-1 hash of the referenced object (ObjectHash)
//
// Entry types by mode:
// - 040000: Directory (tree object)
// - 100644: Regular file (blob object)
// - 100755: Executable file (blob object)
// - 120000: Symbolic link (blob object)
// - 160000: Git submodule (commit object)
//
// Serialized format in tree object:
// [mode] [space] [filename] [null byte] [20-byte SHA-1 binary]
//
// Example serialized entry for "hello.txt" file:
// "100644 hello.txt\0[20 bytes of SHA-1]"
type TreeEntry struct {
	mode objects.FileMode
	name scpath.RelativePath
	sha  objects.ObjectHash
}

// NewTreeEntry creates a new TreeEntry with validation
func NewTreeEntry(mode objects.FileMode, name scpath.RelativePath, sha objects.ObjectHash) (*TreeEntry, error) {
	if !name.IsValid() {
		return nil, fmt.Errorf("invalid path: %s", name)
	}

	if err := sha.Validate(); err != nil {
		return nil, fmt.Errorf("invalid SHA: %w", err)
	}

	entry := &TreeEntry{
		mode: mode,
		name: name.Normalize(),
		sha:  sha,
	}

	return entry, nil
}

// NewTreeEntryFromStrings creates a new TreeEntry from string values (for backward compatibility)
func NewTreeEntryFromStrings(modeStr, name, shaStr string) (*TreeEntry, error) {
	mode, err := objects.FromOctalString(modeStr)
	if err != nil {
		return nil, fmt.Errorf("invalid mode: %w", err)
	}

	path, err := scpath.NewRelativePath(name)
	if err != nil {
		return nil, fmt.Errorf("invalid path: %w", err)
	}

	sha, err := objects.ParseObjectHash(shaStr)
	if err != nil {
		return nil, fmt.Errorf("invalid SHA: %w", err)
	}

	return NewTreeEntry(mode, path, sha)
}

// Mode returns the entry mode
func (e *TreeEntry) Mode() objects.FileMode {
	return e.mode
}

// Name returns the entry name
func (e *TreeEntry) Name() string {
	return e.name.String()
}

// Path returns the entry path
func (e *TreeEntry) Path() scpath.RelativePath {
	return e.name
}

// SHA returns the entry SHA-1 hash
func (e *TreeEntry) SHA() objects.ObjectHash {
	return e.sha
}

// IsDirectory returns true if this entry is a directory
func (e *TreeEntry) IsDirectory() bool {
	return e.mode == objects.FileModeDirectory
}

// IsFile returns true if this entry is a regular or executable file
func (e *TreeEntry) IsFile() bool {
	return e.mode == objects.FileModeRegular || e.mode == objects.FileModeExecutable
}

// IsExecutable returns true if this entry is an executable file
func (e *TreeEntry) IsExecutable() bool {
	return e.mode == objects.FileModeExecutable
}

// IsSymbolicLink returns true if this entry is a symbolic link
func (e *TreeEntry) IsSymbolicLink() bool {
	return e.mode == objects.FileModeSymlink
}

// IsSubmodule returns true if this entry is a submodule
func (e *TreeEntry) IsSubmodule() bool {
	return e.mode == objects.FileModeGitlink
}

// Serialize writes the serialized entry to the provided writer
// Format: [mode] [space] [filename] [null byte] [20-byte SHA-1 binary]
func (e *TreeEntry) Serialize(w io.Writer) error {
	// Write mode, space, and filename
	if _, err := fmt.Fprintf(w, "%s %s%c", e.mode.ToOctalString(), e.name.String(), objects.NullByte); err != nil {
		return fmt.Errorf("failed to write entry header: %w", err)
	}

	// Write SHA bytes
	shaBytes, err := e.sha.Bytes()
	if err != nil {
		return fmt.Errorf("failed to get SHA bytes: %w", err)
	}

	if _, err := w.Write(shaBytes); err != nil {
		return fmt.Errorf("failed to write SHA bytes: %w", err)
	}

	return nil
}

// CompareTo compares this entry with another entry.
// The comparison is based on the entry's name, with directories sorted before files.
// Returns a negative value if this entry is less than the other, zero if they are equal,
// or a positive value if this entry is greater than the other.
func (e *TreeEntry) CompareTo(other *TreeEntry) int {
	if e.name == other.name {
		if e.IsDirectory() && !other.IsDirectory() {
			return -1
		}
		if !e.IsDirectory() && other.IsDirectory() {
			return 1
		}
		return 0
	}
	if e.name < other.name {
		return -1
	}
	return 1
}

// DeserializeTreeEntry creates a TreeEntry from serialized data
func DeserializeTreeEntry(data []byte, offset int) (*TreeEntry, int, error) {
	spaceIndex := bytes.IndexByte(data[offset:], objects.SpaceByte)
	if spaceIndex == -1 {
		return nil, 0, fmt.Errorf("invalid tree entry: missing space")
	}
	spaceIndex += offset

	modeStr := string(data[offset:spaceIndex])

	nullIndex := bytes.IndexByte(data[spaceIndex+1:], objects.NullByte)
	if nullIndex == -1 {
		return nil, 0, fmt.Errorf("invalid tree entry: missing null byte")
	}
	nullIndex += spaceIndex + 1

	nameStr := string(data[spaceIndex+1 : nullIndex])

	// Extract SHA bytes
	start := nullIndex + 1
	end := start + SHALengthBytes
	if end > len(data) {
		return nil, 0, fmt.Errorf("invalid tree entry: incomplete SHA")
	}

	shaBytes := data[start:end]
	shaStr := hex.EncodeToString(shaBytes)

	entry, err := NewTreeEntryFromStrings(modeStr, nameStr, shaStr)
	if err != nil {
		return nil, 0, err
	}

	return entry, end, nil
}
